File: 21-mergeTwoLists.php
<?php

class ListNode {
      public $val = 0;
      public $next = null;
      function __construct($val = 0, $next = null) {
          $this->val = $val;
          $this->next = $next;
      }
}

// 迭代法
function mergeTwoLists ($l1, $l2)
{
    if (!$l1 && !$l2) return null;
    $res = new ListNode();
    $curr = $res; // 当前指针，用来进行偏移

    while ($l1 || $l2) {
        // 如果两个链表都有值
        if (!empty($l1) && !empty($l2)) {
            if ($l1->val < $l2->val) {
                $curr->next = new ListNode($l1->val);
                $l1 = $l1->next;
            } else {
                $curr->next = new ListNode($l2->val);
                $l2 = $l2->next;
            }
            // 执行完要移动当前指针，然后continue不往下执行，否则对后面会有影响
            $curr = $curr->next;
            continue;
        }

        // 如果一个有值，一个没值
        if (empty($l1) && !empty($l2)) {
            $curr->next = new ListNode($l2->val);
            $l2 = $l2->next;
            $curr = $curr->next;
            continue;
        }

        if (!empty($l1) && empty($l2)) {
            $curr->next = new ListNode($l1->val);
            $l1 = $l1->next;
            $curr = $curr->next;
            continue;
        }
    }

    return $res->next;
}

// 优化迭代法
// 时间和空间复杂度都是 O n
function mergeTwoLists11 ($l1, $l2) {
    if (!$l1 && !$l2) return null;

    $res = new ListNode();
    $curr = $res; // 当前指针，用来进行偏移

    // 如果两个链表都不为空，则开始判断
    while ($l1 && $l2) {
        if ($l1->val < $l2->val) {
            $curr->next = $l1;
            $l1 = $l1->next;
        } else {
            $curr->next = $l2;
            $l2 = $l2->next;
        }
        $curr = $curr->next;
    }
    // 如果其中一个链表空了，那么直接取不为空的链表补在后面就行了
    $curr->next = $l1 ?? $l2;
    return $res->next;
}


// 递归法
// 空间复杂度 :  单次执行的空间 * 递归次数 = 1 * n = O(n)
// 时间复杂度 ： 单次执行的时间 * 递归次数 = 1 * n = O(n)
function mergeTwoLists2 ($l1, $l2)
{
    if (!$l1 || !$l2)  return $l1 ?? $l2;

    // 下面需要自己画图才能理解到位
    if ($l1->val < $l2->val) {
        $l1->next = mergeTwoLists2($l1->next, $l2);
        return $l1;
    } else {
        $l2->next = mergeTwoLists2($l1, $l2->next);
        return $l2;
    }
}

File: go-code/21-mergeTwoLists.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy1 := &ListNode{}
	dummy2 := &ListNode{}
	dummyRes := &ListNode{}
	res := dummyRes

	dummy1.Next = list1
	dummy2.Next = list2

	for dummy1.Next != nil && dummy2.Next != nil {
		if dummy1.Next.Val < dummy2.Next.Val {
			dummyRes.Next = &ListNode{Val: dummy1.Next.Val}
			dummy1 = dummy1.Next
		} else {
			dummyRes.Next = &ListNode{Val: dummy2.Next.Val}
			dummy2 = dummy2.Next
		}
		dummyRes = dummyRes.Next
	}

	if dummy1.Next != nil {
		dummyRes.Next = dummy1.Next
	}
	if dummy2.Next != nil {
		dummyRes.Next = dummy2.Next
	}
	return res.Next
}

func mergeTwoListsV2(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil && list2 == nil {
		return nil
	}
	res := &ListNode{}
	curr := res
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			curr.Next = list1
			list1 = list1.Next
		} else {
			curr.Next = list2
			list2 = list2.Next
		}
		curr = curr.Next
	}
	if list1 != nil {
		curr.Next = list1
	}
	if list2 != nil {
		curr.Next = list2
	}

	return res.Next
}
